### PR TITLE
Refactor the API around contexts and modes

### PR DESCRIPTION
**Pull request by chris-wood**
This change decouples the notion of verifiability from the ciphersuite definition, since the algorithms in use do not affect how verifiability works. There are now two "modes" for the OPRF protocol: base mode and verifiability mode. Clients and servers run a "setup" step that produces a client or server context (respectively), which has the relevant APIs on them (blind, unblind, and finalize for clients, and evaluate for servers). The verifiable mode augments (extends) these contexts to provide the additional proof generation and verification logic. 

I also updated the ciphersuite definitions to move H2, which was previously "hash to scalar element," to the GG API. This lets us use hash-to-field from draft-irtf-cfrg-hash-to-curve, which is nice. It will also let us easily adopt ristretto255 (and Decaf), since those have a similar API shape.

cc @armfazh @grittygrease

### CODE REVIEW

File: draft-irtf-cfrg-voprf.md [hunk 1]
```diff
@@ -476,13 +427,20 @@ prime-order group.
   difficult to reverse the mapping. Examples of hash to group functions
   satisfying this property are the ones described for prime-order
   (sub)groups of elliptic curves, see {{I-D.irtf-cfrg-hash-to-curve}}.
+- HashToScalar(): A member function of `GG` that deterministically maps
+  an array of bytes `x` to a random element in the scalar field of `GG`.
+- RandomScalar(): A member function of `GG` that generates a random, non-zero
+  element in the scalar field of `GG`.
+- KeyGen(): A member function of `GG` that generates (skX, pkX), where skX is
+  a random, non-zero element in the scalar field `GG` and pkX is the product
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:435]: I'm don't think that this makes sense as a group API function. There's no concept of a keypair in a group.

File: draft-irtf-cfrg-voprf.md [hunk 2]
```diff
@@ -476,13 +427,20 @@ prime-order group.
...
-Lastly, for any scalar `r` that is an element of the galois field of
-scalars `GF(p)` associated with `GG`, we assume it is always written in
-big-endian byte array format for the purpose of providing wherever it
-is supplied as an input or give as an output of a function.
+<!-- TODO(caw): do we need to define RandomScalar and KeyGen in the ciphersuites section? -->
 
-### Group instantiations
+This document uses the types `Element` and `Scalar` to denote elements of
+the group `GG` and its underlying scalar field, respectively. For notational
+clarity, we use the alias `PublicKey` and `PrivateKey` to refer to items of
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:442]: I think that the `KeyGen` function should be removed from here and we can just talk about these types when we get to the protocol.

File: draft-irtf-cfrg-voprf.md [hunk 3]
```diff
@@ -392,53 +390,6 @@ in the following section.
 
 # Preliminaries
 
-We start by detailing some necessary cryptographic definitions.
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:395]: Is there a reason why this has been removed, is it just not required? FWIW I'm not necessarily against it.

File: draft-irtf-cfrg-voprf.md [hunk 4]
```diff
@@ -505,85 +463,51 @@ multiplication into all elliptic curve operations.
...
-The Server MUST run `KeyGen` to generate `(skS, pkS)`. The
-variable `skS` is used as its private key, and `pkS` is used as
-its public key. Servers that support verifiability MUST make `pkS`
-available to clients.
+The following one-byte values distinguish between these two modes:
 
-## Protocol message flow {#message-flow}
+| Mode            | Value |
+|:================|:======|
+| mode_base       | 0x00  |
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:485]: We seem to have gone back to snake case here :) It seems that a lot of the H2C functions are in snake case, and we use camel case in this document exclusively. Maybe we should make everything snake case instead?

File: draft-irtf-cfrg-voprf.md [hunk 5]
```diff
@@ -505,85 +463,51 @@ multiplication into all elliptic curve operations.
...
-the protocol exchange. Ciphersuites can either be verifiable, or not. A
-verifiable ciphersuite forces the Server to prove the authenticity of
-the evaluation message `ev` that the Server sends using a ZK proof.
+Both participants agree on the mode and a choice of ciphersuite that is used
+before the protocol exchange. Once established, the core protocol works
+as follows:
 
 ~~~
-   Client(ins, pkS, info)                 Server(skS, pkS)
+   Client(input, pkS, info)                 Server(skS, pkS)
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:495]:
```suggestion
   Client(inputs, pkS, info)
```

File: draft-irtf-cfrg-voprf.md [hunk 6]
```diff
@@ -505,85 +463,51 @@ multiplication into all elliptic curve operations.
...
 
-                                    ev = Evaluate(skS, pkS, bts)
+                          evaluation = Evaluate(skS, pkS, bts)
 
-                            ev
+                         evaluation
                         <----------
 
-    unbToks = Unblind(pkS, toks, bts, ev)
+    signedTokens = Unblind(pkS, tokens, blindTokens, evaluation)
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:507]: I tried to deliberately move away from the word "signed" in the past as it sounds like a signature scheme. For Privacy Pass we use "issued", do you think that could work here? I don't have many better suggestions.

File: draft-irtf-cfrg-voprf.md [hunk 7]
```diff
@@ -601,28 +525,50 @@ multiple (V)OPRF applications should use separate auxiliary values to to
 ensure finalized outputs are separate. Guidance for constructing info can
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:525]: Noticed a typo in line 523 above: `in (V)OPRF the protocol` -> `in the (V)OPRF protocol`

File: draft-irtf-cfrg-voprf.md [hunk 8]
```diff
@@ -601,28 +525,50 @@ multiple (V)OPRF applications should use separate auxiliary values to to
...
-## Data structures {#structs}
+## Context Setup
+
+Both modes of the OPRF involve an offline setup phase which establishes a context
+used by the client or server in executing the online phase of the protocol.
+
+~~~
+def SetupBaseServer():
+  (skS, _) = GG.KeyGen()
+  contextString = I2OSP(mode_base, 1) + I2OSP(suite.identifier, 2)
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:536]: Has `suite` been defined somewhere previously?

File: draft-irtf-cfrg-voprf.md [hunk 9]
```diff
@@ -601,28 +525,50 @@ multiple (V)OPRF applications should use separate auxiliary values to to
...
+  return ServerContext(contextString, skS)
+
+def SetupBaseClient(suite):
+  contextString = I2OSP(mode_base, 1) + I2OSP(suite.identifier, 2)
+  return ClientContext(contextString)
+
+def SetupVerifiableServer(suite):
+  (skS, pkS) = GG.KeyGen()
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:544]: Maybe we need to say around here that the server should make this public key public?

File: draft-irtf-cfrg-voprf.md [hunk 10]
```diff
@@ -642,447 +588,327 @@ required.
...
-slightly with the opportunity for making performance gains in some
-scenarios. We detail these modifications in {{blinding}}.
-
-~~~
-Input:
-
- ClientInput inputs[m]
+  PrivateKey skS
+  PublicKey pkS
+  SerializeElement blindedTokens[m]
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:613]:
```suggestion
SerializedElement blindedTokens[m]
```

File: draft-irtf-cfrg-voprf.md [hunk 11]
```diff
@@ -642,447 +588,327 @@ required.
...
 
 Output:
 
- Token tokens[m]
- BlindedToken blindedTokens[m]
-
-Steps:
+  Evaluation ev
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:617]:
```suggestion
  Evaluation Ev
```

File: draft-irtf-cfrg-voprf.md [hunk 12]
```diff
@@ -642,447 +588,327 @@ required.
...
+  for i = 0 to m:
+    di = 1
+    Mi = GG.Deserialize(blindedTokens[i])
+    Zi = GG.Deserialize(ev.elements[i])
+    if m > 1:
+       h2Input = I2OSP(len(seed), 2) || seed || I2OSP(i', 2) ||
+                 I2OSP(len(compositeDST), 2) || compositeDST
+       di = GG.HashToScalar(h2Input)
+
+       if di > GG.order():
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:672]: Is rejection sampling not carried out in `hash_to_field`?

File: draft-irtf-cfrg-voprf.md [hunk 13]
```diff
@@ -642,447 +588,327 @@ required.
...
-    1. BT = GG.Deserialize(blindedTokens[i])
-    2. Z = skS * BT
-    3. elements[i] = GG.Serialize(Z)
- 3. Ev = Evaluation{ elements: elements }
- 4. if verifiable:
-    1. proof = GenerateProof(skS, pkS, blindedTokens, elements)
-    2. Ev.proof = proof
- 5. Output Ev
+def ComputeComposites(gen, pkS, blindedTokens, ev):
+  seedDST = "RFCXXXX-seed" + self.contextString
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:651]: Maybe include an extra hyphen for DSTs: `"RFCXXXX-seed-" + self.contextString`, to make it clear that the final part is a separate identifier?

File: draft-irtf-cfrg-voprf.md [hunk 14]
```diff
@@ -642,447 +588,327 @@ required.
...
+  return (c, s)
 ~~~
 
-### Finalize
+We note here that it is essential that a different r value is used for
+every invocation. If this is not done, then this may leak `skS` in a
+similar fashion as is possible in Schnorr or (EC)DSA scenarios where
+fresh randomness is not used.
+
+#### Evaluate
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:722]: `Evaluate` is the most important algorithm in this section, should we have it above the utility functions above?

File: draft-irtf-cfrg-voprf.md [hunk 15]
```diff
@@ -642,447 +588,327 @@ required.
...
- 4. Output output
+  Evaluation Ev
+
+def Evaluate(skS, pkS, blindedTokens):
+  elements = []
+  for i in 1..m:
+    BT = GG.Deserialize(blindedTokens[i])
+    Z = skS * BT
+    elements[i] = GG.Serialize(Z)
+  proof = GenerateProof(skS, pkS, blindedTokens, ev)
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:741]:
```suggestion
  proof = GenerateProof(skS, pkS, blindedTokens, Ev)
```

File: draft-irtf-cfrg-voprf.md [hunk 16]
```diff
@@ -642,447 +588,327 @@ required.
...
- 3. for i = 0 to m:
-    1. pre = preprocs[i]
-    2. r = pre.blind
-    3. r * G = GG.Deserialize(pre.blindedGenerator)
-    4. P = GG.HashToGroup(inputs[i])
-    5. tokens[i] = Token{ data: x, blind: pre.blindedPublicKey }
-    6. blindedTokens[i] = GG.Serialize(P + r * G)
- 4. Output (tokens, blindedTokens)
-~~~
+This blinding mechanism can be modified slightly with the opportunity
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:779]: This paragraph is already said above and so it can be removed.

File: draft-irtf-cfrg-voprf.md [hunk 17]
```diff
@@ -642,447 +588,327 @@ required.
...
 
- Token tokens[m]
- Evaluation ev
- PublicKey pkS
- BlindedToken blindedTokens[m]
- boolean verifiable
+  PublicKey pkS
+  Token tokens[m]
+  SerializedElement blindedTokens[m]
+  Evaluation ev
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:791]: For consistency:
```suggestion
  Evaluation Ev
```

File: draft-irtf-cfrg-voprf.md [hunk 18]
```diff
@@ -642,447 +588,327 @@ required.
...
 
 Output:
 
- SerializedGroupElement composites[2]
+  SerializedElement unblindedTokens[m]
 
-Steps:
+def Unblind(pkS, tokens, blindedTokens, ev):
+  if VerifyProof(pkS, blindedTokens, ev) == false:
+    ABORT()
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:890]: Is `ABORT()` explained anywhere?

File: draft-irtf-cfrg-voprf.md [hunk 19]
```diff
@@ -642,447 +588,327 @@ required.
...
-In the algorithms below we set the following domain separation labels.
-These are used to control the domains of the hash function evaluations
-that we use.
+  SerializedElement unblindedTokens[m]
 
-~~~
-opaque challengeDST<1..2^16-1>
-opaque seedDST<1..2^16-1>
-opaque compositeDST<1..2^16-1>
+def Unblind(pkS, tokens, blindedTokens, ev):
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:797]:
```suggestion
def Unblind(pkS, tokens, blindedTokens, Ev):
```

File: draft-irtf-cfrg-voprf.md [hunk 20]
```diff
@@ -642,447 +588,327 @@ required.
...
+  unblindedTokens = []
+  for i = 0 to m:
+    r = tokens[i].blind
+    Z = GG.Deserialize(Evaluation.elements[i])
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:801]:
```suggestion
    Z = GG.Deserialize(Ev.elements[i])
```

File: draft-irtf-cfrg-voprf.md [hunk 21]
```diff
@@ -642,447 +588,327 @@ required.
...
 ~~~
 Input:
 
- PublicKey pkS
- BlindedTokens blindedTokens[m]
- Evaluation ev
- Scalar proof[2]
+  PublicKey pkS
+  SerializedElement blindedTokens[m]
+  Evaluation ev
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:847]:
```suggestion
  Evaluation Ev
```

File: draft-irtf-cfrg-voprf.md [hunk 22]
```diff
@@ -642,447 +588,327 @@ required.
...
+  Scalar proof[2]
 
 Output:
 
- boolean verified
+  boolean verified
 
-Steps:
+def VerifyProof(pkS, blindedTokens, ev, proof):
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:854]:
```suggestion
def VerifyProof(pkS, blindedTokens, Ev, proof):
```

File: draft-irtf-cfrg-voprf.md [hunk 23]
```diff
@@ -642,447 +588,327 @@ required.
...
 Input:
 
- SerializedGroupElement gen
- PublicKey pkS
- BlindedTokens blindedTokens[m]
- Evaluation ev
+  PublicKey pkS
+  Token tokens[m]
+  SerializedElement blindedTokens[m]
+  Evaluation ev
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:882]:
```suggestion
  Evaluation Ev
```

File: draft-irtf-cfrg-voprf.md [hunk 24]
```diff
@@ -642,447 +588,327 @@ required.
...
-          1. i = i-1 # decrement and try again
-       4. i  = i + 1
-       5. i' = i' + 1
-    5. M = di * Mi + M
-    6. Z = di * Zi + Z
- 9. Output [GG.Serialize(M), GG.Serialize(Z)]
+  unblindedTokens = []
+  for i = 0 to m:
+    r = tokens[i].blind
+    Z = GG.Deserialize(Evaluation.elements[i])
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:895]:
```suggestion
    Z = GG.Deserialize(Ev.elements[i])
```

File: draft-irtf-cfrg-voprf.md [hunk 25]
```diff
@@ -642,447 +588,327 @@ required.
...
+# Ciphersuites {#ciphersuites}
+
+A ciphersuite for the protocol wraps the functionality required for the
+protocol to take place. This ciphersuite should be available to both
+Client and Server, and agreement on the specific instantiation is
+assumed throughout. A ciphersuite contains instantiations of the
+following functionalities.
+
+- `GG`: A prime-order group exposing the API detailed in {{pog}}.
+- `Hash`: A cryptographic hash function that is indifferentiable from
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:910]: This is a really nice interface! :) I'm glad that we could get rid of the HKDF.

File: draft-irtf-cfrg-voprf.md (hunk 3 above)
Comment (alxdavids) [reply draft-irtf-cfrg-voprf.md:395]: ignore, I see that you've moved it below! :+1

File: draft-irtf-cfrg-voprf.md [hunk 26]
```diff
@@ -642,447 +588,327 @@ required.
...
+  SerializedElement blindedTokens[m]
+  Evaluation ev
 
 Output:
 
- SerializedGroupElement composites[2]
+  SerializedElement unblindedTokens[m]
 
-Steps:
+def Unblind(pkS, tokens, blindedTokens, ev):
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:888]:
```suggestion
def Unblind(pkS, tokens, blindedTokens, Ev):
```

File: draft-irtf-cfrg-voprf.md [hunk 27]
```diff
@@ -642,447 +588,327 @@ required.
...
+  if VerifyProof(pkS, blindedTokens, ev) == false:
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:889]:
```suggestion
  if VerifyProof(pkS, blindedTokens, Ev) == false:
```

**Comment by alxdavids** [comment]
Apologies for the weirdly formatted review comments. I was using the VS code GitHub PR review integration for the first time and it has come out rather strangely.

File: draft-irtf-cfrg-voprf.md (hunk 1 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:435]: Hmm, but there is a notion of a scalar and element of the group, right?

File: draft-irtf-cfrg-voprf.md (hunk 4 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:485]: Good catch! I'll make this snake case.

File: draft-irtf-cfrg-voprf.md (hunk 7 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:525]: Eagle eye!

File: draft-irtf-cfrg-voprf.md (hunk 6 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:507]: I think issued will work. Let's go with that!

File: draft-irtf-cfrg-voprf.md (hunk 9 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:544]: Yep, that'll work.

File: draft-irtf-cfrg-voprf.md (hunk 8 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:536]: I forgot to pass it to this function. Fixed!

File: draft-irtf-cfrg-voprf.md (hunk 14 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:722]: Yep, I'm fine with that. I'll bump it up.

File: draft-irtf-cfrg-voprf.md (hunk 16 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:779]: Good catch!

File: draft-irtf-cfrg-voprf.md (hunk 18 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:890]: It is now. :)

File: draft-irtf-cfrg-voprf.md (hunk 25 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:918]: Me too!

File: draft-irtf-cfrg-voprf.md (hunk 12 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md]: Nope -- it's deterministic.

File: draft-irtf-cfrg-voprf.md (hunk 13 above)
Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md]: Good idea. I'll add that.

**Comment by chris-wood** [comment]
@alxdavids this is ready to go! I'm making it an official PR.

File: draft-irtf-cfrg-voprf.md [hunk 28]
```diff
@@ -650,447 +601,330 @@ required.
...
+  PublicKey pkS
+  SerializedElement blindedTokens[m]
 
 Output:
 
- Token tokens[m]
- BlindedToken blindedTokens[m]
-
-Steps:
+  Evaluation ev
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:634]:
```suggestion
  Evaluation Ev
```

File: draft-irtf-cfrg-voprf.md [hunk 29]
```diff
@@ -650,447 +601,330 @@ required.
...
-modifications in {{blinding}}.
-
-### Evaluate
+### Verifiable Server Context
 
-~~~
-Input:
+The VerifiableServerContext extends the base ServerContext with an augmented
+`Evaluate()` function. This function produces a proof that `skS` was used
+in computing the result. It makes use of the helper functions `ComputeComposites`
```
Comment (alxdavids) [inline draft-irtf-cfrg-voprf.md:650]: Do we need to make these functions snake-case? It looks like they are in sage.

Comment (chris-wood) [reply draft-irtf-cfrg-voprf.md:650]: I think the functions should all consistently use camel-cased. Snake case in the code is just a Python-ism (I think).

**Pull request merged**